ktor-server/ktor-server-core/jvmAndNix/src/io/ktor/server/application/plugins/api/KtorCallContexts.kt

```diff
@@ -14,37 +16,62 @@ import io.ktor.utils.io.*
 
 /**
  * The context associated with the call that is currently being processed by server.
- * Every call handler ([PluginBuilder.onCall], [PluginBuilder.onCallReceive], [PluginBuilder.onCallRespond], and so on)
- * of your plugin has a derivative of [CallHandlingContext] as a receiver.
+ * Every call handler ([ApplicationPluginBuilder.onCall], [ApplicationPluginBuilder.onCallReceive], [ApplicationPluginBuilder.onCallRespond], and so on)
+ * of your plugin has a derivative of [CallContext] as a receiver.
  **/
-public open class CallHandlingContext(internal open val context: PipelineContext<*, ApplicationCall>) {
+@PluginsDslMarker
+public open class CallContext<PluginConfig : Any> internal constructor(
+    public val pluginConfig: PluginConfig,
+    protected open val context: PipelineContext<*, ApplicationCall>
+) {
     // Internal usage for tests only
     internal fun finish() = context.finish()
 }
 
+public open class RestrictedCallContext<PluginConfig : Any> internal constructor(
```

Comment:
Not sure about the name, let's discuss later

Reply:
looks like it can be `sealed class`, no?

Reply:
thanks, will check